link-grammar/connectors.c

```diff
@@ -33,116 +35,221 @@ void free_connectors(Connector *e)
+		econlist = alloca(exp_num_con * sizeof(*econlist));
+		get_connectors_from_expression(econlist, (exp_num_con = 0, &exp_num_con), e);
+	}
+
+	if ((NULL == econlist) || (NULL == econlist[0])) return;
+
+	qsort(econlist, exp_num_con, sizeof(*econlist), condesc_by_uc_num);
+
+	/* Scan the expression connector list and set length_limit.
+	 * restart_cn is needed because several connectors in this list
+	 * may match a given uppercase part. */
+	size_t restart_cn = 0, cn = 0, en;
+	for (en = 0; en < exp_num_con; en++)
+	{
+		for (cn = restart_cn; cn < ct->num_con; cn++)
+			if (sdesc[cn]->uc_num >= econlist[en]->uc_num) break;
+
+		for (; en < exp_num_con; en++)
+			if (econlist[en]->uc_num >= sdesc[cn]->uc_num) break;
+
+		if (econlist[en]->uc_num != sdesc[cn]->uc_num) continue;
+		restart_cn = cn+1;
+
+		const char *wc_str = econlist[en]->string;
+		char *uc_wildcard = strchr(wc_str, WILD_TYPE);
+
+		for (; cn < ct->num_con; cn++)
+		{
+			if (NULL == uc_wildcard)
+			{
+				if (econlist[en]->uc_num != sdesc[cn]->uc_num)
+					break;
+				/* The uppercase parts are equal - match only the lowercase ones. */
+				if (!lc_easy_match(econlist[en], sdesc[cn]))
+					continue;
+			}
+			else
+			{
+				/* The uppercase part is a prefix. */
+				if (0 != strncmp(wc_str, sdesc[cn]->string, uc_wildcard - wc_str))
+					break;
+			}
+
+			sdesc[cn]->length_limit = length_limit;
+		}
+	}
+}
 
-bool match_in_connector_set(Connector_set *conset, Connector * c)
+void set_all_condesc_length_limit(Dictionary dict)
 {
-	unsigned int h;
-	Connector * c1;
-	if (conset == NULL) return false;
-	h = connector_set_hash(conset, c->string, '+');
-	for (c1 = conset->hash_table[h]; c1 != NULL; c1 = c1->next)
+	ConTable *ct = &dict->contable;
+	bool unlimited_len_found = false;
+
+	for (length_limit_def_t *l = ct->length_limit_def; NULL != l; l = l->next)
+	{
+		set_condesc_length_limit(dict, l->defexp, l->length_limit);
+		if (UNLIMITED_LEN == l->length_limit) unlimited_len_found = true;
+	}
+
+	if (!unlimited_len_found)
 	{
-		if (easy_match(c1->string, c->string)) return true;
+		/* If no connectors are defined as UNLIMITED_LEN, set all the
+		 * connectors with no defined length-limit to UNLIMITED_LEN. */
+		condesc_t **sdesc = ct->sdesc;
+
+		for (size_t en = 0; en < ct->num_con; en++)
+		{
+			if (0 == sdesc[en]->length_limit)
+				sdesc[en]->length_limit = UNLIMITED_LEN;
+		}
 	}
-	return false;
+
+	length_limit_def_t *l_next;
+	for (length_limit_def_t *l = ct->length_limit_def; NULL != l; l = l_next)
+	{
+		l_next = l->next;
+		free(l);
+	}
+
+	if (verbosity_level(D_SPEC+1))
+	{
+		printf("%5s %-6s %3s\n", "num", "uc_num", "ll");
+		for (size_t n = 0; n < ct->num_con; n++)
+		{
+			printf("%5zu %6d %3d %s\n", n, ct->sdesc[n]->uc_num,
+			       ct->sdesc[n]->length_limit, ct->sdesc[n]->string);
+		}
+	}
+
 }
 
 /* ======================================================== */
 
+static bool connector_encode_lc(const char *lc_string, condesc_t *desc)
```

Comment:
Can you provide a description of what this does?  It seems to be taking the lower-case part of a connector, normally an 8-bit quantity, and packing it into a 7-bit thing .. that seems like a lot of work, for not much savings.  Since it is lower-case, you could, in principle, pack down to 5 bits, for more savings ... but is it really worth the effort? or am I missing something?

Reply:
OK, never mind. I see what you are doing .. you have to copy the LC chars into the u64 anyway, and you have to compute the mask, anyway, and, so at almost no extra cost, you can shrink to 7 bits.  But then, at almost no extra cost, you can subtract 0x60 and pack in 5 bits. On the other hand, I do not believe we have have more than maybe 5 or 6 lower-case letters... so this is not all that necessary...

Reply:
> But then, at almost no extra cost, you can subtract 0x60 and pack in 5 bits. On the other hand, I do not believe we have have more than maybe 5 or 6 lower-case letters... so this is not all that necessary...

The LC part is not really only lower case letters. It may contain also numbers. So we have 36 possible characters, that can be packed into 6 bits. However, this is more complex than just removing the most significant bit since there is a gape between the values if the lower case letters and the numbers.

Not doing this packing means maximum 8 characters in the LC part, and I agree 9 is not a big improvement and may not even worth the slight overhead of doing it once. The reason I included it is copying from an old code I wrote, that included packing the UC part in 6 bits (the same packing code can be used). Packing the UC part is not needed now because the connector enumeration idea is better.

So I will remove this packing (or ifdef it out).

Also note that all[*] the connector related computations, including LC packing, are now done only once - at the dictionary read time, instead of per sentence (or even per connector) as before.
[*] Besides short_length/all_short_connectors, which can also be computed and cached only when changed, but I didn't implement it.

Reply:
OK. Maybe its not that important.

Reply:
> But then, at almost no extra cost, you can subtract 0x60 and pack in 5 bits. On the other hand, I do not believe we have have more than maybe 5 or 6 lower-case letters... so this is not all that necessary...

With such a sustruction method (e.g substructing 64 and taking the the absolute) it is possible to pack in 6 bits instead of 7 (I was wrong in saying this is more complex due to the gape), and also the longest LC part is 5 - the version (and domain) encoding. So with this packing 32-bits can hold 5 characters. However, this will prevent versions like `V5v4v10`...

BTW, I once hacked the dict reading and added `#define` for dict definitions (a very small change) (`#` is already a special character for the dict).

Another BTW, for not opening an issue for that, I noted that reading the `id` dict fails
 with: `Connectors beginning with "ID" are forbidden`. The easiest fix for that is to forbid only connectors in the format IDX when X is a  capital letter string (as only that may clash with idiom connectors).

Is such a fix is fine?

Reply:
ID: apparently, its needed, so that we can set the indonesian locale. So yes.  

We do need version strings like `V5v4v10` ...

Note that the statistically-generated dictionaries currently use only upper-case letters, and have zero lower-case letters in them. And there are a lot of them:  I haven't counted recently, but maybe hundreds of thousands (or more). The goal of learning is to shrink this number, but ...